### PR TITLE
Fix non required api fields

### PR DESCRIPTION
**Pull request by dbendilas**
Problem
-----------------------
API requests were failing because they included fields that were not expected by the endpoints. Older versions of the API must have been handling extra fields gracefully without complaining, but newer versions return a 400 Bad Response. 

Solution
--------
We now send only those fields that are really expected by each endpoint, without any redundant ones.

This repository hasn't been (really) updated in years, so a major refactoring was also done, together with various changes. In particular:
- Deleted unused configuration parameters, such as `read_only_fields`
- Added a new, convenient way of defining which fields to set when creating or updating an object. 
   Now callers can use: `obj.save(param1=value2, ..., paramN=valueN)` instead of setting each   
   parameter individually and then calling `.save()`
- Removed the custom override in `Resource` for retrieving the contents of the resource, as it 
   conflicted with the new way of defining fields for creation or update. Now the same can be done
   with `resource.get_content()` instead of `resource.content`. 

   The new way seems also more clear, as it explicitly shows that there needs to be a retrieval
   (so a new HTTP request) and the content is not readily available, as it was implied with `.content`.
- Made various formatting changes, for simplicity and readability
- Added docstrings and comments throughout the code
- Added useful information in the README file
- Made various other small changes

How to test
-----------
TODO

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following best practices
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,6 +1,84 @@
-==========================
- Transifex Python Library
-==========================
+Transifex Python Library
+========================
 
-The library is a wrapper around the Transifex API that tries to present
-a simpler interface to the developers.
+The library is a wrapper around the Transifex API, built to provide a
```
Comment (rigaspapas) [inline README.rst]: Maybe you mean "This library"?

File: README.rst [hunk 2]
```diff
@@ -1,6 +1,84 @@
...
+
+Initialization
+~~~~~~~~~~~~~~
+
+Before making any requests, you need to setup the HTTP handler, which
+will be used for all subsequent requests.
+
+Authentication works both with username/password or with an API token.
+
+\```python from txlib.http.auth import BasicAuth from
```
Comment (rigaspapas) [inline README.rst]: Why did you escape the backticks? There is something wrong with the newlines here.

File: txlib/api/base.py [hunk 3]
```diff
@@ -1,45 +1,126 @@
...
+"""
+
+
 import json
 
 from txlib.utils import _logger
 from txlib.registry import registry
 
 
+# noinspection PyPackageRequirements
```
Comment (rigaspapas) [inline txlib/api/base.py]: What is this?

File: txlib/api/base.py [hunk 4]
```diff
@@ -1,45 +1,126 @@
...
+    _is_initialized = False
 
     @classmethod
     def get(cls, **kwargs):
-        """Method to retrieve an object from a Tx server."""
-        url_fields = {}
+        """Retrieve an object by making a GET request to Transifex.
+
+        Each value in `kwargs` that corresponds to a field
+        defined in `self.url_fields` will be used in the URL pth
```
Comment (rigaspapas) [inline txlib/api/base.py]: pth -> path

File: txlib/api/base.py [hunk 5]
```diff
@@ -1,45 +1,126 @@
...
+
+        Raises:
+            txlib.http.exceptions.NotFoundError: if the object with these
+                attributes is not found on the remote server
+            txlib.http.exceptions.ServerError subclass: depending on
+                the particular server response
+
+        Example:
+        # Note: also catch exceptions
+        >>> obj = MyModel.get(attr1=value1, attru2=value2)
```
Comment (rigaspapas) [inline txlib/api/base.py]: There is a typo : `attru2`

File: txlib/api/base.py [hunk 6]
```diff
@@ -1,45 +1,126 @@
...
+        """
+        fields = {}
         for field in cls.url_fields:
-            url_fields[field] = kwargs.pop(field, None)
-        model = cls(**url_fields)
+            fields[field] = kwargs.pop(field, None)
+
+        # Create an instance of the model class and make the GET request
+        model = cls(**fields)
```
Comment (rigaspapas) [inline txlib/api/base.py]: Why create a new instance here? Don't we already have one stored in `cls`?

File: txlib/http/http_requests.py [hunk 7]
```diff
@@ -10,9 +10,7 @@
 class HttpRequest(BaseRequest):
     """Basic http requests handler.
 
-    This class can handle both http and https requests. However, since
-    it uses the `requests` module (which in turn uses `urllib2`), it does
-    not verify the SSL certificate in case of https.
+    This class can handle both http and https requests.
```
Comment (rigaspapas) [inline txlib/http/http_requests.py]: HTTP and HTTPS should always be uppercase I think

File: txlib/api/tests/test_base_model.py [hunk 8]
```diff
@@ -25,3 +34,80 @@ def test_join_subpaths(self):
...
+    def test_setting_invalid_field_raises_error(self):
+        obj = DummyModel(slug='slug')
+        obj.name = 'name'
+        obj.description = 'description'
+        with pytest.raises(AttributeError):
+            obj.invalid_field = ''
+
+    def test_saving_existing_object_updates_it(self):
+        setup_registry(
+            get='{"id": 100, "param": "value"}',
```
Comment (rigaspapas) [inline txlib/api/tests/test_base_model.py]: Are these values used?

File: txlib/api/tests/test_base_model.py [hunk 9]
```diff
@@ -25,3 +34,80 @@ def test_join_subpaths(self):
...
+            put='{"slug": "slug", "name": "name"}',
+        )
+
+        obj = DummyModel.get(something='anything')
+        obj.slug = 'slug'
+        obj.name = 'name'
```
Comment (rigaspapas) [inline txlib/api/tests/test_base_model.py]: Would the test also pass if we didn't set those two values?

File: txlib/api/tests/utils.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    :param str get: the json string to return for all GET requests
+    :param str post: the json string to return for all POST requests
+    :param str put: the json string to return for all PUT requests
+    :param str delete: the json string to return for all DELETE requests
+    """
+    handler = MagicMock()
+    registry.setup({'http_handler': handler})
+
+    if get is not None:
+        registry.http_handler.get.return_value = json.loads(get)
```
Comment (rigaspapas) [inline txlib/api/tests/utils.py]: Why do you use strings as payloads for mocking and the your parse them as JSON? I think we could work directly with dictionaries in the test files.

File: txlib/api/base.py (hunk 6 above)
Comment (dbendilas) [reply txlib/api/base.py]: This is a static method, so `cls` is the class itself, not the instance of the class.

File: txlib/api/tests/test_base_model.py (hunk 9 above)
Comment (dbendilas) [reply txlib/api/tests/test_base_model.py]: No, it fails with `AssertionError: assert None == 'slug'` and `AttributeError: DummyModel has no readable attribute name` respectively,

**Comment by dbendilas** [comment]
@rigaspapas regarding mocking HTTP responses, I totally agree. I had some trouble while mocking on that layer, but now I've managed to make it work.

File: txlib/api/tests/test_base_model.py [hunk 11]
```diff
@@ -15,6 +16,10 @@ class DummyModel(BaseModel):
 class TestBaseModel():
     """Test the base model for the Transifex model wrappers."""
 
+    @pytest.fixture(autouse=True)
+    def auto_clean_registry(self, clean_registry):
```
Comment (rigaspapas) [inline txlib/api/tests/test_base_model.py]: Is this needed/used at all?

Comment (dbendilas) [reply txlib/api/tests/test_base_model.py]: This run for every test (because of `autouse=True`).

Comment (rigaspapas) [reply txlib/api/tests/test_base_model.py]: It does nothing though. I tried removing it and tests still pass. Is this a placeholder for something?
